sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -477,5 +486,82 @@ object StateStore extends Logging {
       None
     }
   }
-}
 
+  private class StateSchemaCompatibilityChecker(
+      providerId: StateStoreProviderId,
+      hadoopConf: Configuration) {
+
+    private val storeCpLocation = providerId.storeId.storeCheckpointLocation()
+    private val fm = CheckpointFileManager.create(storeCpLocation, hadoopConf)
+    private val schemaFileLocation = schemaFile(storeCpLocation)
+
+    fm.mkdirs(schemaFileLocation.getParent)
+
+    def check(keySchema: StructType, valueSchema: StructType): Unit = {
+      if (fm.exists(schemaFileLocation)) {
+        logDebug(s"Schema file for provider $providerId exists. Comparing with provided schema.")
+        val (storedKeySchema, storedValueSchema) = readSchemaFile()
+
+        def typesEq(schema1: StructType, schema2: StructType): Boolean = {
+          val fieldToType: StructField => (DataType, Boolean) = f => (f.dataType, f.nullable)
+          (schema1.length == schema2.length) &&
+            schema1.map(fieldToType).equals(schema2.map(fieldToType))
```

Comment:
Shouldn't be `==` instead of `equals`? Scala does more checks than java.
This question applies to other places too.